### PR TITLE
ddl: fix a bug that we miss adding last handle index in some case.

### PR DESCRIPTION
**Pull request by winkyao**
## What have you changed? (mandatory)

This PR fixes a bug that, when the last remains indices count is just defaultTaskHandleCnt+1, we do not use handleRange.endIncluded outside, so we may miss adding the last index.

In this case, we break fetchRowColVals by `len(w.idxRecords) >= w.batchCnt`, then the handleOutOfRange is false, so the `nextHandle = idxRecords[len(idxRecords)-1].handle + 1`, it is exactly equal to endHandle, so in handleBackfillTask, we break the loop by`finishTask = handleRange.startHandle >= handleRange.endHandle` mistakenly.

master branch will fail the newly added test case with:

```
db_integration_test.go:135:
    tk.MustExec("admin check index t b")
/Users/wink/go_dir/src/github.com/pingcap/tidb/util/testkit/testkit.go:160:
    tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
... value *errors.Err = &errors.Err{message:"", cause:(*errors.Err)(0xc420877130), previous:(*errors.Err)(0xc4208773b0), file:"/Users/wink/go_dir/src/github.com/pingcap/tidb/util/testkit/testkit.go", line:134} ("table count 129 != index(b) count 128")
... sql:admin check index t b, [], error stack /Users/wink/go_dir/src/github.com/pingcap/tidb/util/admin/admin.go:204: table count 129 != index(b) count 128
/Users/wink/go_dir/src/github.com/pingcap/tidb/executor/executor.go:452:
/Users/wink/go_dir/src/github.com/pingcap/tidb/executor/adapter.go:269:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/tidb.go:189:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/session.go:745:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/session.go:808:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/session.go:757:
/Users/wink/go_dir/src/github.com/pingcap/tidb/util/testkit/testkit.go:134:
```

## What is the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

ut

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

NO

## Does this PR affect tidb-ansible update? (mandatory)

NO

## Does this PR need to be added to the release notes? (mandatory)

YES

```
fix a bug that we miss adding last handle index in some case.
```

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -678,7 +684,19 @@ func (w *addIndexWorker) handleBackfillTask(d *ddlCtx, task *reorgIndexTask) *ad
 		}
 
 		handleRange.startHandle = nextHandle
-		if handleRange.startHandle >= handleRange.endHandle {
+		finishTask := false
+		// if nextHandle is math.Int64, we can only use handleOutOfRange to indicates the nexHandle is handled.
```
Comment (zimulala) [inline ddl/index.go]: s/if/If
s/to indicates/to indicate
s/nexHandle/nextHandle

File: ddl/index.go [hunk 2]
```diff
@@ -678,7 +684,19 @@ func (w *addIndexWorker) handleBackfillTask(d *ddlCtx, task *reorgIndexTask) *ad
 		}
 
 		handleRange.startHandle = nextHandle
-		if handleRange.startHandle >= handleRange.endHandle {
+		finishTask := false
```
Comment (lamxTyler) [inline ddl/index.go]: We can move it to line 692, where it is used.

File: ddl/index.go [hunk 3]
```diff
@@ -678,7 +684,19 @@ func (w *addIndexWorker) handleBackfillTask(d *ddlCtx, task *reorgIndexTask) *ad
...
+		// if nextHandle is math.Int64, we can only use handleOutOfRange to indicates the nexHandle is handled.
+		if handleOutOfRange && nextHandle == handleRange.endHandle {
```
Comment (lamxTyler) [inline ddl/index.go]: Is `handleOutOfRange` implies `nextHandle == handleRange.endHandle`?

Comment (winkyao) [reply ddl/index.go]: no. handleOutOfRange just mean current batch is handled out of range.

**Comment by winkyao** [comment]
@zimulala @lamxTyler PTAL

File: ddl/index.go [hunk 4]
```diff
@@ -637,6 +642,7 @@ func (w *addIndexWorker) backfillIndexInTxn(handleRange reorgIndexTask) (nextHan
 
 		if handleOutOfRange || len(idxRecords) == 0 {
 			nextHandle = handleRange.endHandle
+			handleOutOfRange = true
```
Comment (tiancaiamao) [inline ddl/index.go]: I'm confused why `handleOutOfRange = true` when `len(idxRecords) == 0`

Comment (tiancaiamao) [reply ddl/index.go]: And `nextHandle` should not be `handleRange.endHandle` when `handleRange.exclude` is false

**Comment by tiancaiamao** [comment]
PTAL  @shenli

File: ddl/index.go [hunk 5]
```diff
@@ -549,10 +551,26 @@ func (w *addIndexWorker) getIndexRecord(handle int64, recordKey []byte, rawRecor
 	return idxRecord, nil
 }
 
-func (w *addIndexWorker) fetchRowColVals(txn kv.Transaction, taskRange reorgIndexTask) ([]*indexRecord, bool, error) {
+func (w *addIndexWorker) getNextHandle(taskRange reorgIndexTask, handleOutOfRange bool) (nextHandle int64) {
+	nextHandle = taskRange.startHandle
+	if handleOutOfRange {
+		if taskRange.endIncluded && taskRange.endHandle != math.MaxInt64 {
+			nextHandle = taskRange.endHandle + 1
```
Comment (zimulala) [inline ddl/index.go]: Could we add a test case to cover it?

Comment (winkyao) [reply ddl/index.go]: The test case already added in TestMaxHandleAddIndex.

Comment (zimulala) [reply ddl/index.go]: Could we remove this code?

Comment (winkyao) [reply ddl/index.go]: nextHandle is used in handleBackfillTask, even if the handleOutRange is true.

Comment (zimulala) [reply ddl/index.go]: Got it.

**Comment by winkyao** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

**Comment by CaitinChen** [comment]
Here is a suggestion for your PR description:

This PR fixes a bug that when the last remained index count is just `defaultTaskHandleCnt`+1, the last index may not be added because the `handleRange.endIncluded` attribute is not judged in the `handleBackfillTask` function (the caller of `fetchRowColVals`).

In this case, `fetchRowColVals` is broken by `len(w.idxRecords) >= w.batchCnt`, and then `handleOutOfRange` will be false and nextHandle = idxRecords[len(idxRecords)-1].handle + 1. The `nextHandle` value will be exactly equal to the `endHandle` value. Thus in `handleBackfillTask`, the loop is mistakenly broken by `finishTask = handleRange.startHandle >= handleRange.endHandle`.

The newly added test case will fail in the master branch:

```
db_integration_test.go:135:
    tk.MustExec("admin check index t b")
/Users/wink/go_dir/src/github.com/pingcap/tidb/util/testkit/testkit.go:160:
    tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
... value *errors.Err = &errors.Err{message:"", cause:(*errors.Err)(0xc420877130), previous:(*errors.Err)(0xc4208773b0), file:"/Users/wink/go_dir/src/github.com/pingcap/tidb/util/testkit/testkit.go", line:134} ("table count 129 != index(b) count 128")
... sql:admin check index t b, [], error stack /Users/wink/go_dir/src/github.com/pingcap/tidb/util/admin/admin.go:204: table count 129 != index(b) count 128
/Users/wink/go_dir/src/github.com/pingcap/tidb/executor/executor.go:452:
/Users/wink/go_dir/src/github.com/pingcap/tidb/executor/adapter.go:269:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/tidb.go:189:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/session.go:745:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/session.go:808:
/Users/wink/go_dir/src/github.com/pingcap/tidb/session/session.go:757:
/Users/wink/go_dir/src/github.com/pingcap/tidb/util/testkit/testkit.go:134:
```

File: ddl/index.go [hunk 6]
```diff
@@ -549,11 +551,37 @@ func (w *addIndexWorker) getIndexRecord(handle int64, recordKey []byte, rawRecor
 	return idxRecord, nil
 }
 
-func (w *addIndexWorker) fetchRowColVals(txn kv.Transaction, taskRange reorgIndexTask) ([]*indexRecord, bool, error) {
+// getNextHandle gets next handle of entry that we are going to process.
+func (w *addIndexWorker) getNextHandle(taskRange reorgIndexTask, taskDone bool) (nextHandle int64) {
+	if !taskDone {
+		// The task is not done. So we need to pick the last processed entry's handle and plus one.
```
Comment (CaitinChen) [inline ddl/index.go]: The task is not done. So we need to pick the last processed entry's handle and add one.

Note: "plus" is a prep, not a verb.

File: ddl/index.go [hunk 7]
```diff
@@ -549,11 +551,37 @@ func (w *addIndexWorker) getIndexRecord(handle int64, recordKey []byte, rawRecor
...
+		return w.idxRecords[len(w.idxRecords)-1].handle + 1
+	}
+
+	// The task is done. So we need to choose a handle out side this range.
```
Comment (CaitinChen) [inline ddl/index.go]: The task is done. So we need to choose a handle outside this range.

File: ddl/index.go [hunk 8]
```diff
@@ -549,11 +551,37 @@ func (w *addIndexWorker) getIndexRecord(handle int64, recordKey []byte, rawRecor
...
+	// There are some corner cases should be considered:
```
Comment (CaitinChen) [inline ddl/index.go]: Some corner cases should be considered:

Comment (CaitinChen) [reply ddl/index.go]: You can also use "There are some corner cases **which** should be considered:"

**Comment by winkyao** [comment]
@CaitinChen Thanks for your valuable suggestions.

**Comment by CaitinChen** [comment]
@winkyao My pleasure. ^_^

**Pull request merged**